### PR TITLE
Implement mul, floordiv, mod, divmod, and reversed directly in TimedeltaArray

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Companion to #23829, will need rebasing after that goes through.  Some overlap with #23789; this will have a smaller diff after being rebased on top of that.

### CODE REVIEW

File: pandas/core/indexes/datetimelike.py [hunk 1]
```diff
@@ -777,6 +777,12 @@ def wrap_arithmetic_op(self, other, result):
     if result is NotImplemented:
         return NotImplemented
 
+    if isinstance(result, tuple):
+        # divmod, rdivmod
+        assert len(result) == 2
```
Comment (gfyoung) [inline pandas/core/indexes/datetimelike.py]: Would this bubble up in any way?

File: pandas/tests/arithmetic/test_timedelta64.py [hunk 2]
```diff
@@ -1236,6 +1236,66 @@ def test_td64arr_rfloordiv_tdlike_scalar(self, scalar_td, box_with_array):
...
+        tdarr = tm.box_expected(tdi, box_with_array)
+
+        expected = TimedeltaIndex(['1 ns', '0 ns'] * 5)
+        expected = tm.box_expected(expected, box_with_array)
+
+        result = tdarr % 2
+        tm.assert_equal(result, expected)
+
+        with pytest.raises(TypeError):
+            2 % tdarr
```
Comment (gfyoung) [inline pandas/tests/arithmetic/test_timedelta64.py]: Check error message.

File: pandas/tests/arithmetic/test_timedelta64.py [hunk 3]
```diff
@@ -1236,6 +1236,66 @@ def test_td64arr_rfloordiv_tdlike_scalar(self, scalar_td, box_with_array):
...
+        tdarr = tm.box_expected(tdi, box_with_array)
+
+        expected = TimedeltaIndex(['1 Day', '2 Days', '0 Days'] * 3)
+        expected = tm.box_expected(expected, box_with_array)
+
+        result = tdarr % three_days
+        tm.assert_equal(result, expected)
+
+        if box_with_array is pd.DataFrame:
+            pytest.xfail("DataFrame does not have __divmod__ or __rdivmod__")
```
Comment (gfyoung) [inline pandas/tests/arithmetic/test_timedelta64.py]: You have two distinct tests here as delineated by this `xfail` check.  In the interest of modularity, this test should be broken up into IMO.  Similar comment goes for your other tests where there is an `xfail` like this.

**Comment by gfyoung** [comment]
Broader question here: given the amount of moving parts in different PR's, to what extent is your implementations in this PR tested (or will be tested)?  I ask because I see a lot of branching logic, but I'm uncertain as to how many of the branches are covered.

File: pandas/core/indexes/datetimelike.py (hunk 1 above)
Comment (gfyoung) [reply pandas/core/indexes/datetimelike.py]: Referring just to the `len-2`.  While I appreciate dummy-protection, I would still want to know if that `assert` would surface for end-users, that's all.

**Review by jreback — CHANGES_REQUESTED** [review]
same concerns as on other PR

**Comment by jreback** [comment]
can you rebase

**Comment by jbrockmendel** [comment]
rebased+green

File: pandas/core/arrays/timedeltas.py [hunk 4]
```diff
@@ -458,6 +437,155 @@ def __rtruediv__(self, other):
...
+            #  on the i8 values
+            result = other.asi8 // self.asi8
+            mask = self._isnan | other._isnan
+            if mask.any():
+                result = result.astype(np.int64)
+                result[mask] = np.nan
+            return result
+
+        elif is_object_dtype(other):
+            result = [other[n] // self[n] for n in range(len(self))]
```
Comment (jreback) [inline pandas/core/arrays/timedeltas.py]: any nice way to remove some of this duplication (floordir / div), maybe with some helper functions.

File: pandas/core/arrays/timedeltas.py [hunk 5]
```diff
@@ -458,6 +437,155 @@ def __rtruediv__(self, other):
...
+        if isinstance(other, (ABCSeries, ABCDataFrame, ABCIndexClass)):
+            return NotImplemented
+        if isinstance(other, (timedelta, np.timedelta64, Tick)):
+            other = Timedelta(other)
+        return self - (self // other) * other
+
+    def __rmod__(self, other):
+        # Note: This is a naive implementation, can likely be optimized
+        if isinstance(other, (ABCSeries, ABCDataFrame, ABCIndexClass)):
+            return NotImplemented
```
Comment (jreback) [inline pandas/core/arrays/timedeltas.py]: can this share with ``__mod__``?

File: pandas/core/arrays/timedeltas.py [hunk 6]
```diff
@@ -458,6 +437,155 @@ def __rtruediv__(self, other):
...
+        if isinstance(other, (timedelta, np.timedelta64, Tick)):
+            other = Timedelta(other)
+
+        res1 = self // other
+        res2 = self - res1 * other
+        return res1, res2
+
+    def __rdivmod__(self, other):
+        # Note: This is a naive implementation, can likely be optimized
```
Comment (jreback) [inline pandas/core/arrays/timedeltas.py]: can this share with ``__divmod__``?

File: pandas/core/arrays/timedeltas.py [hunk 7]
```diff
@@ -326,37 +301,41 @@ def _addsub_offset_array(self, other, op):
...
+                freq = self.freq * other
+            return type(self)(result, freq=freq)
 
-        return NotImplemented
+        if not hasattr(other, "dtype"):
+            # list, tuple
+            other = np.array(other)
+        if len(other) != len(self) and not is_timedelta64_dtype(other):
+            # Exclude timedelta64 here so we correctly raise TypeError
+            #  for that instead of ValueError
```
Comment (jreback) [inline pandas/core/arrays/timedeltas.py]: lots of this impl matches how this is done in Timedelta. too bad can't easily share.

File: pandas/core/arrays/timedeltas.py [hunk 8]
```diff
@@ -458,6 +437,155 @@ def __rtruediv__(self, other):
...
+            result = self._data // other
+            return type(self)(result)
+
+        else:
+            dtype = getattr(other, "dtype", type(other).__name__)
+            raise TypeError("Cannot divide {typ} by {cls}"
+                            .format(typ=dtype, cls=type(self).__name__))
+
+    def __rfloordiv__(self, other):
+        if isinstance(other, (ABCSeries, ABCDataFrame, ABCIndexClass)):
```
Comment (jreback) [inline pandas/core/arrays/timedeltas.py]: can this share with floordiv?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/timedeltas.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: Not really.  For div and divmod either timedeltas or numeric are valid.  For reversed ops only timedeltas are valid.

File: pandas/core/arrays/timedeltas.py (hunk 7 above)
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: After DTA/TDA are done I might look at this more seriously.

File: pandas/core/arrays/timedeltas.py [hunk 9]
```diff
@@ -332,37 +307,41 @@ def _addsub_offset_array(self, other, op):
...
+        if not hasattr(other, "dtype"):
+            # list, tuple
+            other = np.array(other)
+        if len(other) != len(self) and not is_timedelta64_dtype(other):
+            # Exclude timedelta64 here so we correctly raise TypeError
+            #  for that instead of ValueError
+            raise ValueError("Cannot multiply with unequal lengths")
+
+        if is_object_dtype(other):
+            # this multiplication will succeed only if all elements of other
```
Comment (jreback) [inline pandas/core/arrays/timedeltas.py]: might be useful to put this in a method that can be shared across multiple ops

File: pandas/core/arrays/timedeltas.py [hunk 10]
```diff
@@ -464,6 +443,155 @@ def __rtruediv__(self, other):
...
+            # list, tuple
+            other = np.array(other)
+        if len(other) != len(self):
+            raise ValueError("Cannot divide with unequal lengths")
+
+        elif is_timedelta64_dtype(other):
+            other = type(self)(other)
+
+            # numpy timedelta64 does not natively support floordiv, so operate
+            #  on the i8 values
```
Comment (jreback) [inline pandas/core/arrays/timedeltas.py]: this looks an awful lot like _maybe_mask_missing (in Index)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/timedeltas.py [hunk 11]
```diff
@@ -332,37 +307,41 @@ def _addsub_offset_array(self, other, op):
             raise TypeError("Cannot add/subtract non-tick DateOffset to {cls}"
                             .format(cls=type(self).__name__))
 
-    def _evaluate_with_timedelta_like(self, other, op):
-        if isinstance(other, ABCSeries):
-            # GH#19042
+    def __mul__(self, other):
+        other = lib.item_from_zerodim(other)
```
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/arrays/timedeltas.py]: right, but now you are missing some code here on several of the operators (from_zero_dim), so maybe better to fix now.

**Review by jreback — APPROVED** [review]

**Comment by jbrockmendel** [comment]
> let's really really try to limit code adding.

Sounds good.  This was the last PR I had for TimedeltaArray.  The only one I have left for DatetimeArray is to implement DatetimeArray._from_sequence.  Following that, all I have on my plate is a) to_datetime/array_to_datetime improvements, b) parametrizing+de-duplicating arithmetic tests, and c) DTA/TDA/PA reductions.  All three can wait until after 0.24.0.

**Pull request merged**